README.md
# TokenD Kotlin wallet

This library implements transactions and keys management for TokenD-related projects.

## Installation

For **Gradle** add following lines to your project's `build.gradle`:
```groovy
allprojects {
    repositories {
        ...
        maven { url "https://maven.tokend.io" }
    }
}

dependencies {
    ...
    compile "org.tokend:wallet:3.7.0"
}

```

## Usage examples

Key management and signing:

```kotlin
val SEED = "SBUFJEEK7FMWXPE4HGOWQZPHZ4V5TFKGSF664RAGT24NS662MKTQ7J6S".toCharArray()
private val DATA = "TokenD is awesome".toByteArray()

val account = Account.fromSecretSeed(SEED)
val decoratedSignature = account.signDecorated(DATA)
```

Transaction creation:

```kotlin
val SEED = "SBUFJEEK7FMWXPE4HGOWQZPHZ4V5TFKGSF664RAGT24NS662MKTQ7J6S".toCharArray()
val NETWORK = NetworkParams("Example Test Network")

val sourceAccount = Account.fromSecretSeed(SEED)
val operation = CreateBalanceOp(SOURCE_ACCOUNT_ID, "OLE")

val transaction = TransactionBuilder(NETWORK, sourceAccount.accountId)
                    .addOperation(Operation.OperationBody.ManageBalance(operation))
                    .setMemo(Memo.MemoText("TokenD is awesome"))
                    .addSigner(sourceAccount)
                    .build()

val envelope = transaction.getEnvelope().toBase64()
```

Decoding:
```kotlin
val txResultEncoded = "AAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAJAAAAAAAAAADMOKDasWPzpJIqN9sWipdvcjEZRTnGBvUezXbEd6rKMAAAAAAAAAAA"
val txResult = TransactionResult.fromBase64(txResultEncoded)
```

## ProGuard
As long as you don't decode XDRs (`.fromXdr(...)`) no extra ProGuard
rules are required. Otherwise add following lines to your
project's `proguard-rules.pro`:
```proguard
# Wallet
-keep class org.tokend.wallet.xdr.* { *; }
-keepattributes RuntimeVisibleAnnotations, RuntimeVisibleParameterAnnotations
```
## XDR Update
XDR sources are located in [TokenD XDR repository](https://github.com/tokend/xdr/).
You can generate new XDRs using our Docker-based XDR generator.
[Docker](https://www.docker.com/) is required to perform this action.

In order to generate new XDRs run `generateXDR` script with a source revision (tag or branch or commit) as an argument:

```bash
./generateXDR.sh master
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Please check our [developers guide](https://gitlab.com/tokend/developers-guide)
for further information about branching and tagging conventions.

## [3.7.0] 2021-11-25

### Added
- `Destroyable` implementation in `Account`
- `equal` and `hashCode` methods for `Account`

### Changed
- From now `Account` is always a complete keypair with a private key.
To verify signatures use `Account.verifySignature` static methods
- Updated XDR version to `7708446fd03153bf0d99b299c9df054dd9788c7c`
- Updated Kotlin version to 1.4.10

### Fixed
- Inability to convert true `UInt64` amounts from and to precised

### Removed
- Positive number restriction on transaction salt
- `canSign` method from account (see above)
- `Account.fromAccountId` and `Account.fromPublicKey` methods (see above)

## [3.6.5] 2020-08-12

### Changed
- Updated XDR version to `d639694`

## [3.6.4] 2020-07-15

### Changed
- XDR strings are now encoded end decoded in UTF-8

## [3.6.3] 2020-01-20

### Added
- `Serializable` marker to `NetworkParams`

## [3.6.2] 2019-12-25

### Fixed
- Incorrect salt transform in `Transaction` constructor, use
true absolute value now

## [3.6.1] 2019-11-26

### Added
- `Transaction` constructor from XDR `TransactionEnvelope`

### Changed
- Signing-related `Transaction` methods are now static

## [3.6.0] 2019-11-05

### Changed
- Updated XDR version to `78afc23`

## [3.5.0] 2019-09-24

### Changed
- Updated XDR version to `bfc2e7b`
- Updated TokenD Maven repo domain

## [3.4.2] 2019-09-06

### Fixed
- XDR decoding on devices running Java 7

### Changed
- ProGuard rules

## [3.4.1] 2019-09-05

### Changed
- ProGuard rules

## [3.4.0] 2019-09-04

### Added
- Ability to decode XDR models: call `.fromXdr` or `.fromBase64` methods
of the required class in Kotlin or use `*.Decoder` static member in Java
- ProGuard rules
- Ability to add a collection of operation bodies to
`TransactionBuilder`

### Removed
- Apache encoding libraries

### Changed
- Updated XDR version to `9199f20`

## [3.3.0] 2019-07-15

### Added
- Ability to add `DecoratedSignature` to the transaction directly

## [3.2.0] 2019-06-18

### Changed
- Updated XDR version to `cd889b0`

## [3.1.0] 2019-05-27

### Added
- Ability to add transaction signers with the `TransactionBuilder`

### Changed
- Updated way of XDR generation, check out Readme
- Updated XDR version to `3.3.0` (`c8561fd`)

## [3.0.1] 2019-02-28

### Added
- `reference` optional param for `SimplePaymentOp`
- Account rules and roles
- Signer rules and roles

### Changed
- XDR version to `7e06563`
- `PaymentV2` and all related `-V2` things to just `Payment`

### Fixed
- Wrong value type for `KeyValueEntryValue`

[3.0.1]: https://github.com/tokend/kotlin-wallet/compare/1.0.13...3.0.1
[3.1.0]: https://github.com/tokend/kotlin-wallet/compare/3.0.1...3.1.0
[3.2.0]: https://github.com/tokend/kotlin-wallet/compare/3.1.0...3.2.0
[3.3.0]: https://github.com/tokend/kotlin-wallet/compare/3.2.0...3.3.0
[3.4.0]: https://github.com/tokend/kotlin-wallet/compare/3.3.0...3.4.0
[3.4.1]: https://github.com/tokend/kotlin-wallet/compare/3.4.0...3.4.1
[3.4.2]: https://github.com/tokend/kotlin-wallet/compare/3.4.1...3.4.2
[3.5.0]: https://github.com/tokend/kotlin-wallet/compare/3.4.2...3.5.0
[3.6.0]: https://github.com/tokend/kotlin-wallet/compare/3.5.0...3.6.0
[3.6.1]: https://github.com/tokend/kotlin-wallet/compare/3.6.0...3.6.1
[3.6.2]: https://github.com/tokend/kotlin-wallet/compare/3.6.1...3.6.2
[3.6.3]: https://github.com/tokend/kotlin-wallet/compare/3.6.2...3.6.3
[3.6.4]: https://github.com/tokend/kotlin-wallet/compare/3.6.3...3.6.4
[3.6.5]: https://github.com/tokend/kotlin-wallet/compare/3.6.4...3.6.5
[3.7.0]: https://github.com/tokend/kotlin-wallet/compare/3.6.5...3.7.0
[Unreleased]: https://github.com/tokend/kotlin-wallet/compare/3.7.0...HEAD


src/main/kotlin/org/tokend/wallet/utils/Base64.kt
package org.tokend.wallet.utils

import java.io.ByteArrayOutputStream

object Base64 {
    private val encodeTable = (CharRange('A', 'Z') + CharRange('a', 'z') + CharRange('0', '9') + '+' + '/').toCharArray()

    private val decodeTable = intArrayOf(-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1,
            -1, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
            -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1)

    fun encode(data: ByteArray): ByteArray {
        val output = ByteArrayOutputStream()
        var padding = 0
        var position = 0
        while (position < data.size) {
            var b = data[position].toInt() and 0xFF shl 16 and 0xFFFFFF
            if (position + 1 < data.size) b = b or (data[position + 1].toInt() and 0xFF shl 8) else padding++
            if (position + 2 < data.size) b = b or (data[position + 2].toInt() and 0xFF) else padding++
            for (i in 0 until 4 - padding) {
                val c = b and 0xFC0000 shr 18
                output.write(encodeTable[c].toInt())
                b = b shl 6
            }
            position += 3
        }
        for (i in 0 until padding) {
            output.write('='.toInt())
        }
        return output.toByteArray()
    }

    fun decode(encoded: ByteArray): ByteArray {
        val output = ByteArrayOutputStream()
        var position = 0
        while (position < encoded.size) {
            var b: Int
            if (decodeTable[encoded[position].toInt()] != -1) {
                b = decodeTable[encoded[position].toInt()] and 0xFF shl 18
            } else {
                position++
                continue
            }
            var count = 0
            if (position + 1 < encoded.size && decodeTable[encoded[position + 1].toInt()] != -1) {
                b = b or (decodeTable[encoded[position + 1].toInt()] and 0xFF shl 12)
                count++
            }
            if (position + 2 < encoded.size && decodeTable[encoded[position + 2].toInt()] != -1) {
                b = b or (decodeTable[encoded[position + 2].toInt()] and 0xFF shl 6)
                count++
            }
            if (position + 3 < encoded.size && decodeTable[encoded[position + 3].toInt()] != -1) {
                b = b or (decodeTable[encoded[position + 3].toInt()] and 0xFF)
                count++
            }
            while (count > 0) {
                val c = b and 0xFF0000 shr 16
                output.write(c.toChar().toInt())
                b = b shl 8
                count--
            }
            position += 4
        }
        return output.toByteArray()
    }
}